README.md
# Pitcher Cherry Release
Command line utility to help with deploying commits to customer enviornments by doing cherry-picks.
While we know cherry-picking for release is not the best way to do it, it is required in some cases. This package makes the process of creating release feature-branches and cherry-picks faster, easier and less error prone.

## Usage

You can execute cherry-release iinside any git repository with `npx`. Use `--help` to find out about the options
```
npx @pitcher/cherry-release --help
```

```
Usage: cherry-release [options]

Options:
  -t, --targets <branch-names...>  Where do you want to deploy? Provide target release branch names.
                                   You can provide many branch names separated by spaces.
  -b, --branch-name <branch-name>  What should be the unique branch name of your deploy branch(es)? Provide the core of the release feature
                                   branch name that will be used with this template `deploy/<branch-name>_to_<target>`. (default: "2022-12-2")
  -c, --commits <commits...>       What commits do you want to deploy? Provide space-separated SHAs of the commits you want to cherry-pick.
                                   If both commits and search options are provided, they will be merged.
  -s, --search <string>            What JIRA issues do you want to deploy? Specify keyword to search for in the commit messages and
                                   interactively pick them from the list. Case insensitive.
  -np, --no-push                   You don't want to push yet? Disable pushing new branches to the remote repository.
  -h, --help                       display help for command
```


CHANGELOG.md
# [1.4.0](https://github.com/PitcherAG/pitcher-cherry-release/compare/v1.3.1...v1.4.0) (2022-12-16)


### Features

* add support for repos cloned with https ([315227a](https://github.com/PitcherAG/pitcher-cherry-release/commit/315227a704cd9ec60258ca14091277dece3e7bfe))
* allow to pick from 30 most recent commits if neither search nor commits parameters are provided ([39820ca](https://github.com/PitcherAG/pitcher-cherry-release/commit/39820ca8ed64a87aba7f044dc926d72d77d7a51b))
* make merge commits more distinctive ([3eabe02](https://github.com/PitcherAG/pitcher-cherry-release/commit/3eabe02824b289b348ff2ede4c2bcce36eb3767d))

## [1.3.1](https://github.com/PitcherAG/pitcher-cherry-release/compare/v1.3.0...v1.3.1) (2022-12-02)


### Bug Fixes

* typo ([456aebe](https://github.com/PitcherAG/pitcher-cherry-release/commit/456aebe1b2cde2def791fb0c9e79576afff6044e))

# [1.3.0](https://github.com/PitcherAG/pitcher-cherry-release/compare/v1.2.0...v1.3.0) (2022-12-01)


### Features

* brand new command-line interface ([75adef6](https://github.com/PitcherAG/pitcher-cherry-release/commit/75adef69577ddee376ad8bc45d4ad3acd359d611))

# [1.2.0](https://github.com/PitcherAG/pitcher-cherry-release/compare/v1.1.1...v1.2.0) (2022-12-01)


### Bug Fixes

* add jest.setup.js ([b73bf2d](https://github.com/PitcherAG/pitcher-cherry-release/commit/b73bf2df93b783232fc7efeeba64b2904b404583))


### Features

* typescript ([c812f52](https://github.com/PitcherAG/pitcher-cherry-release/commit/c812f5250da38c28f4f4518f6b46e5e372620c14))

## [1.1.1](https://github.com/PitcherAG/pitcher-cherry-release/compare/v1.1.0...v1.1.1) (2022-11-25)


### Bug Fixes

* add new line to output ([e5ff56f](https://github.com/PitcherAG/pitcher-cherry-release/commit/e5ff56ffc8a3e57b441a858b89b1491dfddcff22))

# [1.1.0](https://github.com/PitcherAG/pitcher-cherry-release/compare/v1.0.1...v1.1.0) (2022-11-25)


### Features

* print less empty space ([4f67af1](https://github.com/PitcherAG/pitcher-cherry-release/commit/4f67af1a6ccc25a85da16399c76c51c36a4f7cfd))

## [1.0.1](https://github.com/PitcherAG/pitcher-cherry-release/compare/v1.0.0...v1.0.1) (2022-11-25)


### Bug Fixes

* add missing packages ([3bc69b9](https://github.com/PitcherAG/pitcher-cherry-release/commit/3bc69b929f31e3d17006e49f0f6f1c491595497f))

# 1.0.0 (2022-11-25)


### Features

* initial version ([2686363](https://github.com/PitcherAG/pitcher-cherry-release/commit/26863632496f239ff4cb9bae6e14ce406af855d8))


src/main.spec.ts
describe('Foo', () => {
  it('bar', () => {
      expect(true).toEqual(true)
  })
})


tests/example.test.js
const init = require('../src/init')
const mockExit = jest.spyOn(process, 'exit').mockImplementation(() => {})

describe('example test', () => {
  it('test', () => {
    expect(true).toBe(true)
  })
})


tests/helpers/setup.js
process.argv.push('--fileID=123456')
process.argv.push('--platform=ios')


release.config.js
module.exports = {
  branches: [{ name: 'main' }],
  plugins: [
    '@semantic-release/commit-analyzer',
    '@semantic-release/release-notes-generator',
    ['@semantic-release/changelog', { changelogFile: 'CHANGELOG.md' }],
    ['@semantic-release/npm', { npmPublish: true, pkgRoot: '.' }],
    [
      '@semantic-release/github',
      {
        successComment: false,
        failComment: false,
        releasedLabels: [
          'released<%= nextRelease.channel ? ` on ${nextRelease.channel}@${nextRelease.version}` : "" %> from <%= branch.name %>',
        ],
      },
    ],
    [
      '@semantic-release/git',
      {
        assets: ['dist/pitcher-cherry-release.js', 'CHANGELOG.md', 'package.json'],
        message: 'chore(release): set `package.json` to ${nextRelease.version} [skip ci]\n\n${nextRelease.notes}',
      },
    ],
  ],
}


src/cherry-release.mjs
import inquirer from 'inquirer';
import { execSync } from 'node:child_process'
import chalk from 'chalk'

const sha1 = /\b[0-9a-f]{5,40}\b/g
const todayDate = new Date()
const today = `${todayDate.getFullYear()}-${todayDate.getMonth() + 1}-${todayDate.getDate()}`

function exec(command, silent) {
  if (!silent) console.log(chalk.green('Executing: ') + command)
  const result = execSync(command, { encoding: 'utf8' })
  if (!silent) console.log(result)
  return result
}

inquirer
  .prompt([
    {
      type: 'checkbox',
      name: 'targets',
      message: 'Where do you want to deploy?',
      choices: [
        'sandoz/test',
        'sandoz/qa',
        'sandoz/qa-markets',
        'training',
        'prod',
      ],
    },
    {
      type: 'input',
      name: 'commits',
      message: "What commits do you want to deploy?",
      validate(input) {
        if (sha1.test(input)) {
          return true;
        }

        throw Error('Please provide a valid git commit(s) SHA.');
      },
      filter(input) {
        return input.match(sha1)
      }
    },{
      type: 'input',
      name: 'branch',
      message: 'What should be the unique branch name of your deploy branch(es)?',
      default: today,
      transformer(answer) {
        return `deploy/${answer || today}_to_<env>`
      },
    },
  ])
  .then((answers) => {
    const prUrls = []
    const repo = exec('git config --get remote.origin.url', true).trim().split('.git')[0].split(':')[1]
    const initialBranch = exec('git rev-parse --abbrev-ref HEAD', true).trim()

    const isRepoClean = exec('git status', true).trim().includes('nothing to commit')
    if (!isRepoClean) {
      console.log(chalk.red('You have uncommitted changes. Please stash them before continuing to ensure you don\'t loose any work.'))
      return
    }

    answers.targets.forEach((target) => {
      const featureBranch = `deploy/${answers.branch}_to_${target}`

      exec(`git checkout ${target}`)
      exec(`git pull`)
      exec(`git checkout ${featureBranch}`)
      answers.commits.forEach((commitHash) => {
        exec(`git cherry-pick ${commitHash}`)
      })
      exec(`git push -u origin ${featureBranch}`)

      prUrls.push(`https://github.com/${repo}/compare/${target}...${featureBranch}`)
    })

    console.log('\n\n')
    console.log(chalk.green('You can create your release PRs (CMD + Click to open):'))
    console.log(chalk.green('==================================='))
    prUrls.forEach((url) => {
      console.log(chalk.blue(url))
    })
    console.log('\n\n')

    exec(`git checkout ${initialBranch}`, true)
  })
  .catch((error) => {
    if (error.isTtyError) {
      // Prompt couldn't be rendered in the current environment
    } else {
      console.log(chalk.red('Error!'))
      console.log(chalk.red('==================================='))
      console.log(error)
    }
  });

src/main.ts
#!/usr/bin/env node

import inquirer from 'inquirer'
import { execSync } from 'node:child_process'
import chalk from 'chalk'
import { Command } from 'commander'

const program = new Command()
const sha1 = /\b[0-9a-f]{5,40}\b/g
const todayDate = new Date()
const today = `${todayDate.getFullYear()}-${todayDate.getMonth() + 1}-${todayDate.getDate()}`

function exec(command: string, silent: boolean = false) {
  if (!silent) {
    console.log(chalk.green('Executing: ') + command)
  }
  const result = execSync(command, { encoding: 'utf8' })
  if (!silent) {
    console.log(result)
  }
  return result.trim()
}

function execSilently(command: string) {
  return exec(command, true)
}

function checkPreconditions() {
  const isGitRepo = execSilently('git rev-parse --is-inside-work-tree').includes('true')
  if (!isGitRepo) {
    console.log(chalk.red('This directory is not a git repository.'))
    process.exit()
  }

  const isRepoClean = execSilently('git status').includes('nothing to commit')
  if (!isRepoClean) {
    console.log(chalk.red('You have uncommitted changes. Please stash them before continuing to ensure you don\'t loose any work.'))
    process.exit()
  }
}

(async () => {
  program
    .requiredOption('-t, --targets <branch-names...>', 'Where do you want to deploy? Provide target release branch names.')
    .requiredOption('-b, --branch-name <branch-name>', 'What should be the unique branch name of your deploy branch(es)? Provide the core of the release feature branch name that will be used with this template `deploy/<branch-name>_to_<target>`.', today)
    .option('-c, --commits <commits...>', 'What commits do you want to deploy? Provide commit SHAs that you want to cherry-pick. If both commits and search options are provided, they will be merged.', [])
    .option('-s, --search <string>', 'What JIRA issues do you want to deploy? Specify keyword to search for in the commit messages and interactively pick them from the list. Case insensitive.', '')
    .option('-np, --no-push', 'You don\'t want to push yet? Disable pushing new branches to the remote repository.')
    .parse()

  const options = program.opts()
  checkPreconditions()

  const showRecentCommitsByUsingEmptySearch = options.commits.length === 0 && options.search === ''
  if (options.search || showRecentCommitsByUsingEmptySearch) {
    const commitList = execSilently(`git log -i --grep="${options.search}" --pretty="format:%h<|>%ad<|>%aN<|>%s" --date=format-local:"%Y-%m-%d %H:%M:%S" --max-count=30`)
      .split('\n')
      .map(line => {
        const [sha, date, author, message] = line.split('<|>')
        return { sha, date, author, message }
      })

    if (commitList.length > 0) {
      const inquiry = await inquirer
        .prompt([
          {
            type: 'checkbox',
            name: 'commits',
            message: 'Which of the matching commits from the current branch do you want to deploy?',
            loop: false,
            pageSize: 20,
            choices: commitList.map((c) => {
              const isMergeCommit = c.message.startsWith('Merge')
              const name = isMergeCommit
                ? `${chalk.grey(c.sha)} ${chalk.grey(c.date)}  ${chalk.grey(c.message)}  ${chalk.grey(c.author)}`
                : `${chalk.yellow(c.sha)} ${chalk.magenta(c.date)}  ${c.message}  ${chalk.green(c.author)}`
              return {
                value: c.sha,
                name,
                disabled: isMergeCommit,
              }
          }),
          },
        ])
      options.commits = [
        ...options.commits,
        ...inquiry.commits.reverse() // we want the oldest commits first
      ]
    }
  }

  if (options.commits.length === 0) {
    console.log(chalk.red('No commits provided to be released. Either use -c to provide SHAs to cherry-pick or -s <search> to manually select commits matching criteria.'))
    process.exit()
  }

  console.log('Options: ', program.opts())


  const prUrls: string[] = []
  const originUrl = execSilently('git config --get remote.origin.url')
  const repo = originUrl
    .split('.git')[0]
    .split(originUrl.startsWith('https://') ? 'https://github.com/' : ':')[1]
  const initialBranch = execSilently('git rev-parse --abbrev-ref HEAD')

  options.targets.forEach((target: string) => {
    const featureBranch = `deploy/${options.branchName}_to_${target}`

    exec(`git checkout ${target}`)
    exec(`git pull`)
    exec(`git switch -c ${featureBranch}`)
    options.commits.forEach((commitHash: string) => {
      exec(`git cherry-pick ${commitHash}`)
    })
    if (options.push) {
      exec(`git push -u origin ${featureBranch}`)
    }

    prUrls.push(`https://github.com/${repo}/compare/${target}...${featureBranch}`)
  })

  console.log('\n\n')
  console.log(chalk.green('You can create your release PRs (CMD + Click to open):'))
  console.log(chalk.green('==================================='))
  prUrls.forEach((url) => {
    console.log(chalk.blue(url))
  })
  console.log('\n\n')

  execSilently(`git checkout ${initialBranch}`)
})()